pkg/core/statesync/module.go

```diff
@@ -0,0 +1,469 @@
+	blocksSynced
+)
+
+// Module represents state sync module and aimed to gather state-related data to
+// perform an atomic state jump.
+type Module struct {
+	lock sync.RWMutex
+	log  *zap.Logger
+
+	// syncPoint is the state synchronisation point P we're currently working against.
+	syncPoint uint32
+	// syncStage is the stage of the sync process.
+	syncStage stateSyncStage
+	// syncInterval is the delta between two adjacent state sync points.
+	syncInterval uint32
+	// blockHeight is the index of the latest stored block.
+	blockHeight uint32
+
+	dao     *dao.Simple
+	bc      blockchainer.Blockchainer
+	mptpool *Pool
+
+	billet *mpt.Billet
+}
+
+// NewModule returns new instance of statesync module.
+func NewModule(bc blockchainer.Blockchainer, log *zap.Logger, s *dao.Simple) *Module {
+	if !(bc.GetConfig().P2PStateExchangeExtensions && bc.GetConfig().RemoveUntraceableBlocks) {
+		return &Module{
+			dao:       s,
+			bc:        bc,
+			syncStage: inactive,
+		}
+	}
+	return &Module{
+		dao:          s,
+		bc:           bc,
+		log:          log,
+		syncInterval: uint32(bc.GetConfig().StateSyncInterval),
+		mptpool:      NewPool(),
+		syncStage:    none,
+	}
+}
+
+// Init initializes state sync module for the current chain's height with given
+// callback for MPT nodes requests.
+func (s *Module) Init(currChainHeight uint32) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.syncStage != none {
+		return nil
+	}
+
+	p := (currChainHeight / s.syncInterval) * s.syncInterval
+	if p < 2*s.syncInterval {
+		// chain is too low to start state exchange process, use the standard sync mechanism
+		s.syncStage = inactive
+		return nil
+	}
+	pOld, err := s.dao.GetStateSyncPoint()
+	if err == nil && pOld >= p-s.syncInterval {
+		// old point is still valid, so try to resync states for this point.
+		p = pOld
+	} else {
+		if s.bc.BlockHeight() > p-2*s.syncInterval {
+			// chain has already been synchronised up to old state sync point and regular blocks processing was started
+			s.syncStage = inactive
+			return nil
+		}
+		// There's no old sync point or an old point is still there, but it is outdated.
+		// Also, either
+		//    1) blocks processing wasn't yet started (i.e. state sync process
+		//       wasn't completed for the outdated state sync point)
+		// or 2) blocks processing was started, but latest stored block is outdated
+		//       to continue state sync process.
+		// As far as we can't ruin current chain's state until new state is completely fetched,
+		// outdated blocks, AppExecResults and transfers will be removed from storage during
+		// (*Blockchain).JumpToState(...) execution. And right here we need to clean outdated MPT state.
+		err = s.bc.GetStateModule().CleanStorage()
+		if err != nil {
+			return fmt.Errorf("failed to remove outdated MPT data from storage: %w", err)
+		}
+		// TODO: natives won't be able to initialize cache after this before the state sync end
+		err = s.dao.DeleteStorageItems()
```

Comment:
But why? You only touch storage space after full MPT sync (and block fetching), so you have full MPT now and all you need to do is just so "unwrap" it into plain storage items. The node can be killed while doing this, but it should be able to recover by just completing the process (it's not hard to organize that). You never need to delete these items though.

Reply:
Given https://github.com/nspcc-dev/neo-go/pull/2019#discussion_r693318565 there might be some outdated items in the storage after the previous node run. And it might happen that the current node state does not contain these items. So we need to remove them in order not to mess up the contracts with these outdated items.

Note, that current sync module implementation performs "unwrapping" on-the-fly, so storage item is saved in the storage as soon as the corresponding MPT leaf is received via P2P. Should we change this behaviour and perform one more MPT traverse to store all leaves in one batch after full MPT sync?

Reply:
And I got the idea of https://github.com/nspcc-dev/neo-go/pull/2019#discussion_r692414684, we just don't need to support such situations. We should start state sync process only from genesis or from the previous uncompleted state sync point, right?

In this case, if we're starting with an outdated state sync point stored in the DB, then storage may still contain outdated storage items because of the `on-the-fly unwrapping`. Then this behaviour should be changed.

Reply:
OK, if we're saving storage items on the go then it makes some sense. Technically it should be a little faster this way, so let's keep it for now.

Reply:
Finally, I fixed this. Now we're only doing the jump once. Sync process can be started either from the genesis block or from a valid previously uncompleted state sync point. Other options require dropping the DB manually (a proper error is returned to the user).

Also, I removed `DeleteStorageItems` method. We don't need to remove storage items in described scenarios, because all we have before the state sync process can be started is storage items of the genesis block created by natives initialisation. These items will be overwritten by the new ones during MPT sync process.

Reply:
There's one potential problem connected with this:
- Imagine that node with clear DB is joining the network and should collect state for the state sync point `P`. Imagine also that there's a native contract `X` that was initialized in the genesis block. During initialization `X` puts storage item `Y` in the storage. Genesis block is processed by our node and state sync process for `P` height is started.
- Imagine that contract `X` *removes* item `Y` from the storage in block `P-1`, so that there's no corresponding LeafNode in the `P`-th MPT state.
- Our node successfully completes state sync process and continues normal blocks processing. Imagine that contract `X` tries to get item `Y` from the storage in `P+1`-th block.

Here's the problem: the initial value of `Y` item will be returned from the storage by our node, while other nodes will return nothing. It happens because we don't remove outdated storage items during state sync process.

But it's an artificial situation, currently all our native contracts *do not completely remove initial items from the storage*. Initial items can only be overwritten in next blocks.

Reply:
This problem actually suggests that we should be flushing MPT to storage items in a single pass. Until the jump is made the node technically still has state corresponding to the genesis block and it can be overwritten while we're receiving MPT which is not correct. We can play with temporary prefixes though (flush new values to another prefix and then just rename in a single pass).

Reply:
Fixed. Now we store MPT leaves on-the-fly under `STTempStorage` prefix. Once state sync data collected items with `STStorage` prefix will be replaced with fresh ones.